### PR TITLE
Add support arduino portenta h7

### PR DESCRIPTION
**Pull request by benjaminbjornsson**

### CODE REVIEW

File: boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7.dts
Comment (erwango) [inline boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7.dts:19]: ?

File: boards/arm/arduino_portenta_h7/doc/index.rst
Comment (erwango) [inline boards/arm/arduino_portenta_h7/doc/index.rst:93]: Please provide more detailed instructions.
As this is DFU this is not as direct as flashing using openocd/ST-Link.

File: boards/arm/arduino_portenta_h7/doc/index.rst
Comment (erwango) [inline boards/arm/arduino_portenta_h7/doc/index.rst:128]: Is it accurate?

File: boards/arm/arduino_portenta_h7/Kconfig.defconfig
Comment (erwango) [inline boards/arm/arduino_portenta_h7/Kconfig.defconfig:23]: Please remove as not enabled in dts

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for this work.
This looks globally fine, but some clean up required

File: boards/arm/arduino_portenta_h7/board.c
Comment (benjaminbjornsson) [reply boards/arm/arduino_portenta_h7/board.c:20]: I've chosen to turn led1 off here since I've experienced that the Arduino bootloader leaves it on.
Arduino does the same in the variantInit function found in variants.cpp.

File: boards/arm/arduino_portenta_h7/Kconfig.board
Comment (benjaminbjornsson) [reply boards/arm/arduino_portenta_h7/Kconfig.board:10]: Shouldn't be there, thanks!

File: boards/arm/arduino_portenta_h7/doc/index.rst
Comment (benjaminbjornsson) [reply boards/arm/arduino_portenta_h7/doc/index.rst:93]: Should of course add instructions about double tapping before flashing. But other than that I'm just using west flash as always. Should I add instructions about installing dfu-utils or just mention that it's required?

File: boards/arm/arduino_portenta_h7/doc/index.rst
Comment (benjaminbjornsson) [reply boards/arm/arduino_portenta_h7/doc/index.rst:128]: No that should not be there, thanks!

File: boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7.dts
Comment (benjaminbjornsson) [reply boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7.dts:19]: I'm not sure what you're referring to?

File: boards/arm/arduino_portenta_h7/arduino_portenta_h7-common.dtsi
Comment (benjaminbjornsson) [reply boards/arm/arduino_portenta_h7/arduino_portenta_h7-common.dtsi:42]: Good question, what would make it required?
It should maybe be disabled by default?

File: boards/arm/arduino_portenta_h7/doc/index.rst
Comment (erwango) [reply boards/arm/arduino_portenta_h7/doc/index.rst]: > Should I add instructions about installing dfu-utils or just mention that it's required?

Not required. As dfu is already supported by west, one should be able to find dfu instructions as part of west doc (at least I hope so).

File: boards/arm/arduino_portenta_h7/arduino_portenta_h7-common.dtsi
Comment (erwango) [reply boards/arm/arduino_portenta_h7/arduino_portenta_h7-common.dtsi]: > Good question, what would make it required?

It would be required if used as a source by an IP (but his is not possible yet).

> It should maybe be disabled by default?

It's already disabled by default in *.dtsi

File: boards/arm/arduino_portenta_h7/board.c
Comment (erwango) [reply boards/arm/arduino_portenta_h7/board.c:20]: Ok, then please add a comment about it.

File: boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7.dts
Comment (erwango) [reply boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7.dts:19]: `zephyr,bt-uart = &usart1;` That's a BLE specific property

**Comment by romainreignier** [comment]
Thnaks @benjaminbjornsson for this work!
I have started to do the same last month, but because I am new to Zephyr, the path was a bit difficult for me.

File: boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7.dts
Comment (romainreignier) [inline boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7.dts:53]: Is there a reason to not set the clock to its maximal value of 480 MHz?

File: boards/arm/arduino_portenta_h7/arduino_portenta_h7-common.dtsi
Comment (romainreignier) [inline boards/arm/arduino_portenta_h7/arduino_portenta_h7-common.dtsi:121]: I am not very confident, but it seems to me that only the [stm32h743.dtsi](https://github.com/zephyrproject-rtos/zephyr/blob/7fe5ab315a4453322780e5655364fe4de2a1b9b6/dts/arm/st/h7/stm32h743.dtsi#L22) file has a definition for the `usbotg_fs`.

On the Portenta, as I have understood the schematic, only the High Speed interface is available without modifying R44 and R45 on the board:
image

That's why I have opened this issue: https://github.com/zephyrproject-rtos/zephyr/issues/45094

File: boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7.yaml
Comment (romainreignier) [inline boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7.yaml:14]: I am not sure what this line means but maybe an `arduino_i2c` could also be added?

And, should not these lines added in the dts?
```
arduino_spi: &spi2 {};
arduino_i2c: &i2c3 {};
arduino_serial: &usart1 {};
```

File: boards/arm/arduino_portenta_h7/arduino_portenta_h7-common.dtsi
Comment (benjaminbjornsson) [reply boards/arm/arduino_portenta_h7/arduino_portenta_h7-common.dtsi:121]: To my understanding, the USB HS is the one connected to the boards USB-C connector. This goes through an external PHY which is not yet supported in Zephyr. The USB FS on the other hand is reachable on the Portenta Breakout board, which is what I used to verify the usb/console sample.

Comment (romainreignier) [reply boards/arm/arduino_portenta_h7/arduino_portenta_h7-common.dtsi:121]: Ok, I see.
Is it ok to define it here while only available with an extension board?

File: dts/arm/st/h7/stm32h7.dtsi
Comment (romainreignier) [inline dts/arm/st/h7/stm32h7.dtsi:896]: Maybe the definition in `[stm32h743.dtsi](https://github.com/zephyrproject-rtos/zephyr/blob/7fe5ab315a4453322780e5655364fe4de2a1b9b6/dts/arm/st/h7/stm32h743.dtsi)` should be removed then?
https://github.com/zephyrproject-rtos/zephyr/blob/7fe5ab315a4453322780e5655364fe4de2a1b9b6/dts/arm/st/h7/stm32h743.dtsi#L22-L34

I have noticed that the interrupts are not in the same order and h743 version has only 3 interrupts.

File: boards/arm/arduino_portenta_h7/arduino_portenta_h7-common.dtsi
Comment (romainreignier) [inline boards/arm/arduino_portenta_h7/arduino_portenta_h7-common.dtsi:12]: Which button is referred here?
I have checked the schematics and for me, the button on the Portenta is directly connected to `NRST` and the one on the Breakout Board is connected to the PMIC `PWRON`.

**Comment by romainreignier** [comment]
I have tried your branch myself and managed to run some samples:

- basic/blinky on M7
- basic/blinky on **M4**
- hello_world on M7
- subsys/console/echo on M7
- subsys/shell/shell_module on M7
- drivers/adc on M7

For the adc sample, I had to add these lines to the common dtsi:
```
&adc1 {
	pinctrl-0 = <&adc1_inp0_pa0>;
	pinctrl-names = "default";
	status = "okay";
};
```
And this overlay in the sample:
```
/ {
	zephyr,user {
		io-channels = <&adc1 0>;
	};
};
```

I don't own the Breakout Board to try the USB FS but I have a MachineControl on which I might test it.
I also have the Ethernet Vision shield on which I plan to test the Ethernet.

Everything went well :+1:

File: boards/arm/arduino_portenta_h7/arduino_portenta_h7-common.dtsi
Comment (romainreignier) [reply boards/arm/arduino_portenta_h7/arduino_portenta_h7-common.dtsi:121]: I have tried the usb/console sample on the MachineControl (which is using USB FS like the Breakout board) but the ACM port does not show up on Linux.

File: dts/arm/st/h7/stm32h7.dtsi
Comment (erwango) [reply dts/arm/st/h7/stm32h7.dtsi:896]: @romainreignier Thanks for raising this.
Actually this change is not correct. Only stm32H742 "and higher" support usbotg_fs. On the other hand usbotg_hs is indeed supported on the whole stm32h7 series.

File: boards/arm/arduino_portenta_h7/arduino_portenta_h7_m4.yaml
Comment (erwango) [inline boards/arm/arduino_portenta_h7/arduino_portenta_h7_m4.yaml:12]: `arduino_gpio` represents the usual arduino uno connector which is not present here. Please remove.

File: boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7.yaml
Comment (erwango) [reply boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7.yaml:14]: This should be removed as arduino Uno connector is not available.
Same for `arduino_gpio`

File: boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7_defconfig
Comment (erwango) [inline boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7_defconfig:28]: Uncommon, is that made on purpose ?

**Review by erwango — CHANGES_REQUESTED** [review]

File: boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7_defconfig
Comment (romainreignier) [reply boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7_defconfig:28]: From my understanding of [this doc](https://docs.zephyrproject.org/latest/develop/flash_debug/host-tools.html#sam-boot-assistant-sam-ba) (*Without ROM bootloader" tab), this board has a bootloader at the start of the flash so this config allows to set the partition table in the device tree.

File: boards/arm/arduino_portenta_h7/arduino_portenta_h7_m4.yaml
Comment (romainreignier) [reply boards/arm/arduino_portenta_h7/arduino_portenta_h7_m4.yaml:12]: For the longer term, a arduino_mkr connecter device tree binding could be created but not needed for this PR.

File: boards/arm/arduino_portenta_h7/board.c
Comment (erwango) [inline boards/arm/arduino_portenta_h7/board.c:8]:
```suggestion
#include <zephyr/init.h>
#include <zephyr/drivers/gpio.h>
```
Cf https://github.com/zephyrproject-rtos/zephyr/pull/45417

**Review by erwango — CHANGES_REQUESTED** [review]

File: dts/arm/st/h7/stm32h7.dtsi
Comment (benjaminbjornsson) [reply dts/arm/st/h7/stm32h7.dtsi:896]: Great catch! I'll move it to stm32h745.dtsi

File: boards/arm/arduino_portenta_h7/arduino_portenta_h7-common.dtsi
Comment (benjaminbjornsson) [reply boards/arm/arduino_portenta_h7/arduino_portenta_h7-common.dtsi:12]: You're right, this should be removed.

File: boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7_defconfig
Comment (erwango) [inline boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7_defconfig:1]: Please add copyright

**Review by erwango — CHANGES_REQUESTED** [review]
Few last comment on my side.

Also, can you comment on the core frequency question ?

File: dts/arm/st/h7/stm32h7.dtsi
Comment (romainreignier) [reply dts/arm/st/h7/stm32h7.dtsi]: I am still unable to use the sample `samples/subsys/usb/console` on the MachineControl, the USB device is not detected by the kernel on boot.

File: boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7.dts
Comment (benjaminbjornsson) [reply boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7.dts:52]: I took a look at this again which raised an issue. When trying to run the board at a higher frequency the UART gets corrupted.

My guess is that it has to do with the 25 MHz oscillator and that it's somehow inactive? I added the oscen regulator to the DTS to pull the OSCEN high during boot but is this enough?
I also just noted that in the Arduino source code they write to the PMIC over I2C to activate the 3V1 rail, which is the voltage source for the oscillator. I have tried to write to the same address in another SYS_INIT but the UART is still corrupted at higher frequencies...

File: boards/arm/arduino_portenta_h7/arduino_portenta_h7-common.dtsi
Comment (benjaminbjornsson) [reply boards/arm/arduino_portenta_h7/arduino_portenta_h7-common.dtsi:112]: Have you looked at the output from the UART while running this sample? I've successfully been running both subsys/usb/cdc_acm and subsys/usb/console on my Mac.

Comment (romainreignier) [reply boards/arm/arduino_portenta_h7/arduino_portenta_h7-common.dtsi:112]: Sorry, I have misunderstood the MachineControl schematic, only the USB A socket is USB FS, the micro-usb is USB HS.
I could run the subsys/usb/cdc_acm and subsys/usb/console samples on my side, sorry about that.

File: boards/arm/arduino_portenta_h7/doc/index.rst
Comment (romainreignier) [inline boards/arm/arduino_portenta_h7/doc/index.rst:20]: There is no user button on this board

**Comment by romainreignier** [comment]
If I recap, the missing points on this PR are:
* A copyright
* Understand the core frequency issue
Is that all?

**Comment by benjaminbjornsson** [comment]
> If I recap, the missing points on this PR are:
> 
> * A copyright
> * Understand the core frequency issue
>   Is that all?

Yes, I fixed the other comments locally and since I couldn’t resolve the frequency problem I didn’t push it. I’d like to solve it but I spent too many hours without progress on this…

**Comment by romainreignier** [comment]
Ok, thanks for the update and the hard work! :+1: 
I might try as well.
Is there a mean to see at which frequency the processor is actually running?

From what I have read in the ArduinoCore code, the PMIC is configured by the bootloader. In the Arduino init code, they only reduce the current of the 3.1 V rail back to 1.2 A instead of 2 A used for the Wifi startup. It should not hurt to let it to 2A for now.

Also, from [this forum thread](https://forum.arduino.cc/t/portenta-h7-stm32cubeide-please-join-me/1010091), a guy managed to use its Portenta from CubeMX, even without the bootloader so we might some info about the clocks from this project.

**Comment by romainreignier** [comment]
Indeed, I have tried the `hello_world` example with this value and it does not work:
```
&pll {
	div-m = <5>;
	mul-n = <192>;
	div-p = <2>;
	div-q = <2>;
	div-r = <2>;
	clocks = <&clk_hse>;
	status = "okay";
};

&rcc {
	clocks = <&pll>;
	clock-frequency = <DT_FREQ_M(480)>;
};
```

But interestingly, when running the blinky sample, the pulse duration is 4,80 seconds instead of 1 second. So there is a serious clock issue in this configuration.

capture_blink_arduino

**Comment by romainreignier** [comment]
I have tried to mimic this part of the ArduinoCore code:
https://github.com/arduino/mbed-os/blob/fb9e38d41f947ac9844dd877a1b8e613bbea1eec/targets/TARGET_STM/TARGET_STM32H7/TARGET_STM32H747xI/TARGET_PORTENTA_H7/system_clock_override.c#L132-L141

with:
```c
	LL_AHB4_GRP1_EnableClock(LL_AHB4_GRP1_PERIPH_GPIOH);
	LL_GPIO_SetPinMode(GPIOH, LL_GPIO_PIN_1, LL_GPIO_MODE_OUTPUT);
	LL_GPIO_SetPinOutputType(GPIOH, LL_GPIO_PIN_1, LL_GPIO_OUTPUT_PUSHPULL);
	LL_GPIO_SetPinSpeed(GPIOH, LL_GPIO_PIN_1, LL_GPIO_SPEED_FREQ_LOW);
	LL_GPIO_SetPinPull(GPIOH, LL_GPIO_PIN_1, LL_GPIO_PULL_UP);
	int wait_loop_counter = 1000;
	while(wait_loop_counter != 0)
	{
		wait_loop_counter--;
	}
	LL_GPIO_SetOutputPin(GPIOH, LL_GPIO_PIN_1);
```
Here: https://github.com/zephyrproject-rtos/zephyr/blob/d2a792dd39e588b41a0834db066b7ba387122132/drivers/clock_control/clock_stm32_ll_h7.c#L727

But nothing has changed.

So the OSCEN part should be ok. It must be a config issue somewhere.

@erwango can you spot something in the dts files?

**Comment by romainreignier** [comment]
I have noticed that the following line might be needed in `arduino_portenta_h7_m7_defconfig`:
```
CONFIG_POWER_SUPPLY_SMPS_1V8_SUPPLIES_LDO=y
```
But it does not help by itself :(

**Comment by benjaminbjornsson** [comment]
@romainreignier thanks for testing on your side! I’m starting to think that the best way forward may be to let the board run on the internal clock and let a future PR solve the clock issue?

**Comment by romainreignier** [comment]
Yes, if it is acceptable by the maintainers, we could stay with this clock config for now. But make it clear in the documentation.

**Comment by romainreignier** [comment]
Great, so now, the device tree needs to be modified to use HSI, right?

File: boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7.dts
Comment (galak) [inline boards/arm/arduino_portenta_h7/arduino_portenta_h7_m7.dts:27]: remove label property.

File: dts/arm/st/h7/stm32h745.dtsi
Comment (galak) [inline dts/arm/st/h7/stm32h745.dtsi:50]: remove `label` property.

File: dts/arm/st/h7/stm32h745.dtsi
Comment (galak) [inline dts/arm/st/h7/stm32h745.dtsi:96]: remove `label` property.

**Review by galak — DISMISSED** [review]
remove `label` properties that got commented on from devicetree.

**Comment by benjaminbjornsson** [comment]
I'll try to have this PR updated this weekend.

File: boards/arm/arduino_portenta_h7/doc/index.rst
Comment (erwango) [reply boards/arm/arduino_portenta_h7/doc/index.rst:20]: @benjaminbjornsson Can you check this ?

**Comment by erwango** [comment]
@benjaminbjornsson can you check the build issues reported by CI ?

**Comment by benjaminbjornsson** [comment]
> @benjaminbjornsson can you check the build issues reported by CI ?

@erwango I'm confused, it's not the board being added that fails now. Should more boards be excluded from `flash_shell` or did I break something by adding the new board to the exclude list?

**Comment by erwango** [comment]
> > @benjaminbjornsson can you check the build issues reported by CI ?
> 
> @erwango I'm confused, it's not the board being added that fails now. Should more boards be excluded from `flash_shell` or did I break something by adding the new board to the exclude list?

ok, please rebase and force push

**Comment by erwango** [comment]
@benjaminbjornsson It seems that you're out of luck and hit be several issues non related with your commit.
2 of them were unknown till now, and I raised dedicated issues (see https://github.com/zephyrproject-rtos/zephyr/issues/48911 and https://github.com/zephyrproject-rtos/zephyr/issues/48912).
Other one is known (https://github.com/zephyrproject-rtos/zephyr/issues/48608) but seems to be a tricky one.

Meanwhile, can you check following remaning comment: https://github.com/zephyrproject-rtos/zephyr/pull/45221#discussion_r923043484 ?

**Comment by romainreignier** [comment]
It seems that the only remaining issue is the line about the user button in the description text file.
It should not take too much time to remove @benjaminbjornsson 
Thank you.

**Comment by erwango** [comment]
> It seems that the only remaining issue is the line about the user button in the description text file. It should not take too much time to remove @benjaminbjornsson Thank you.

PR also requires a rebase.
Note that we're now in the process of releasing V3.2.0 and PR that are not bug fixes will now have to wait that the V3.2.0 is fully released before being merged.

**Comment by benjaminbjornsson** [comment]
Sorry for the delay, I've removed the line about the push-button in the index now and also rebased.

**Review by erwango — APPROVED** [review]

**Review by FRASTM — APPROVED** [review]

**Pull request merged**
